glib2: Split package into separate library packages, to cut down on size

pnd10 opened the pull request:
Maintainer: Peter Wagner <tripolar@gmx.at>
Compile tested: ath79, TPlink TL-WDR3600, r24166+6-53039bf7f5
Run tested: ath79, TPlink TL-WDR3600, r24166+6-53039bf7f5, tested by installation and the fact that the gkrellmd package pulls in fewer libraries and the image is smaller
Description: In order to be able to reduce the installed size of anything that depends of GLib2.0, which has 5 component libraries, split the package into 5 separate packages, and a full version which depends on on the others to pull in all five. Also test the split by altering gkrellmd to only pull in the specific libraries that it actually depends on

pnd10 committed:
glib2: split package into separate libraries

Existing package contains five libraries, some of which are > 1MB. Allow
packages which depend on only some of them to only pull in the ones they
need by splitting this package into finer-grained packages. Transition
can be done piecemeal as depending on glib2 will still pull in all the
libraries.

Signed-off-by: Peter Denison <openwrt@marshadder.org>

BKPepe commented:
CI is failing, but I am thinking about one thing related to your PR, which bothers me a lot.

**Why?** You are saying to cut down on size. **What is the difference? Does it make any sense? Is it worth doing it after so many years like it is?** Every year, the demand for HW resources is higher and higher. It is same for computer games.

pnd10 commented:
On Mon, 30 Oct 2023, Josef Schlehofer wrote:

> CI is failing, but I am thinking about one thing related to your PR, which
> bothers me a lot.
> 
> Why? You are saying to cut down on size. What is the difference? Does it
> make any sense? Is it worth doing it after so many years like it is? Every
> year, the demand for HW resources is higher and higher. It is same for
> computer games.

Thanks for looking at it. Yes, I couldn't work out why the CI is failing.

The main reason I did it is that I updated, built exactly the same 
configuration as last time, and it wouldn't fit into my device 
(TL-WDR3600) any more. I went looking for the biggest things and found 
that GLib dwarfed everything else, and 2 out of the 5 libraries I didn't 
need at all, they just came packaged along with everything.

This potentially saves 2-3MB, pre-compression, which is significant when 
the device only has 8MB flash.

This should be a fairly non-invasive change. Anything that depends on 
glib2 will still pull in all the libraries, albeit with 6 packages, not 
one, but packages can if they only need some of the libraries only depend 
on the individual libraries, as I've demonstrated wth gkrellmd.

Works for me!

Peter

jmv2009 commented:
Uhhh, don't you need to figure out then which packages need what exactly? Unless you assign the full version to all of them...

pnd10 commented:
> Uhhh, don't you need to figure out then which packages need what exactly? Unless you assign the full version to all of them...

Yes, ultimately. This can go in, first, without breaking anything, then packages that depend on glib2 can gradually be moved across to the fine-grained packages.

BKPepe commented:
Let's go back on track. Multiple CI checks are failing.

1. Formality one:
```
: === Checking commit '891ff7292b103fb5c06d8f9299db9703fa20ce5f'
: Author name (Peter Denison) seems ok
: Commit subject line MUST start with '<package name>: ' (Split glib2 package into separate library packages, to cut down on size)
: Signed-off-by is missing or doesn't match author (should be 'Signed-off-by: Peter Denison <openwrt@marshadder.org>')
```

2. Compile ones:
```
Package glib2-gio is missing dependencies for the following libraries:
libgmodule-2.0.so.0
libgobject-2.0.so.0
make[2]: *** [Makefile:213: /builder/bin/packages/arm_cortex-a15_neon-vfpv4/packages_ci/glib2-gio_2.74.0-5_arm_cortex-a15_neon-vfpv4.ipk] Error 1
make[2]: Leaving directory '/feed/libs/glib2'
time: package/feeds/packages_ci/glib2/compile#0.47#0.30#0.62
    ERROR: package/feeds/packages_ci/glib2 failed to build.
```

pnd10 commented:
> Let's go back on track. Multiple CI checks are failing.
>     1. Formality one:
> ```
> : Commit subject line MUST start with '<package name>: ' (Split glib2 package into separate library packages, to cut down on size)
> : Signed-off-by is missing or doesn't match author (should be 'Signed-off-by: Peter Denison <openwrt@marshadder.org>')
> ```
Apologies for this, should be easy to fix - shall I redo the commit and do a new PR? Or can this be fixed in this PR?

>     2. Compile ones:
> ```
> Package glib2-gio is missing dependencies for the following libraries:
> libgmodule-2.0.so.0
> libgobject-2.0.so.0
> make[2]: *** [Makefile:213: /builder/bin/packages/arm_cortex-a15_neon-vfpv4/packages_ci/glib2-gio_2.74.0-5_arm_cortex-a15_neon-vfpv4.ipk] Error 1
> make[2]: Leaving directory '/feed/libs/glib2'
> time: package/feeds/packages_ci/glib2/compile#0.47#0.30#0.62
>     ERROR: package/feeds/packages_ci/glib2 failed to build.
> ```
Gah - that's a typo - I'll fix that when I fix the above.

BKPepe commented:
There is no reason to open a new PR. You can do it within existing one.

Dne po 30. 10. 2023 23:29 uživatel Peter Denison ***@***.***>
napsal:

> Let's go back on track. Multiple CI checks are failing.
> 1. Formality one:
>
> : Commit subject line MUST start with '<package name>: ' (Split glib2 package into separate library packages, to cut down on size)
> : Signed-off-by is missing or doesn't match author (should be 'Signed-off-by: Peter Denison ***@***.***>')
>
> Apologies for this, should be easy to fix - shall I redo the commit and do
> a new PR? Or can this be fixed in this PR?
>
> 2. Compile ones:
>
> Package glib2-gio is missing dependencies for the following libraries:
> libgmodule-2.0.so.0
> libgobject-2.0.so.0
> make[2]: *** [Makefile:213: /builder/bin/packages/arm_cortex-a15_neon-vfpv4/packages_ci/glib2-gio_2.74.0-5_arm_cortex-a15_neon-vfpv4.ipk] Error 1
> make[2]: Leaving directory '/feed/libs/glib2'
> time: package/feeds/packages_ci/glib2/compile#0.47#0.30#0.62
>     ERROR: package/feeds/packages_ci/glib2 failed to build.
>
> Gah - that's a typo - I'll fix that when I fix the above.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/packages/pull/22544#issuecomment-1786146872>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA7IDVFXX3VOEZ5LAJB5C5TYCAS5LAVCNFSM6AAAAAA6VC24R2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTOOBWGE2DMOBXGI>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

stangri commented:
@BKPepe this PR modifies Makefiles for both admin/gkrellmd and lib/glib2. Would it not make sense to split into two separate commits, better yet PRs?

pnd10 commented:
> @BKPepe this PR modifies Makefiles for both admin/gkrellmd and lib/glib2. Would it not make sense to split into two separate commits, better yet PRs?

I can certainly split into two commits if you think it would be clearer. I thought putting in one PR would make sense since the change to admin/gkrellmd is a template following the split on how to make use of the newly split packages. I'm happy to go with the consensus though.

pnd10 committed:
gkrellmd: use finer-grained dependencies on glib2

Following the split of glib2, depend on the more fine-grained packages
to reduce the installed size. Gkrellmd only needs to depend on the core glib2
library, gmodule and gthread, and not on gobject or gio.

Signed-off-by: Peter Denison <openwrt@marshadder.org>

GeorgeSapkin commented:
@pnd10 no updates in over two years, still relevant?

pnd10 commented:
> @pnd10 no updates in over two years, still relevant?

Yes, I think it still is. I addressed all of the review comments, and it all went quiet. What do I have to do to actually have someone merge it, or should I just do it myself?

pprindeville commented on admin/gkrellmd/Makefile:
```diff
@@ -26,7 +26,7 @@ include $(INCLUDE_DIR)/nls.mk
   SECTION:=admin
   CATEGORY:=Administration
-  DEPENDS:=+glib2
+  DEPENDS:=+glib2-core +glib2-gmodule +glib2-gthread
```
Version bump needed?

pnd10 replied:
Would you like me to rebase and roll the version bump in, or are you happy with an add-on review commit (presumably one per package, so they can meet the commit format reqs)?

pprindeville replied:
Roll it in and rebase.  Thanks

pprindeville commented on libs/glib2/Makefile:
```diff
@@ -33,14 +33,68 @@ include $(INCLUDE_DIR)/package.mk
 include $(INCLUDE_DIR)/nls.mk
```
Version bump needed?

pprindeville reviewed (changes requested):
Needs `PKG_RELEASE` bumps but other than that looks okay.

pprindeville reviewed (approved):
LGTM

pnd10 commented:
Everything ready. Can someone merge this now?

BKPepe commented:
This causes serious failures on OpenWrt buildbots:
- https://downloads.openwrt.org/snapshots/faillogs/aarch64_cortex-a53/packages/libqmi/compile.txt
```
Run-time dependency mbim-glib found: NO (tried pkgconfig and cmake)

../../../../build_dir/target-aarch64_cortex-a53_musl/libqmi-1.38.0/meson.build:212:18: ERROR: Dependency "mbim-glib" not found, tried pkgconfig and cmake

A full log can be found at /builder/shared-workdir/build/sdk/build_dir/target-aarch64_cortex-a53_musl/libqmi-1.38.0/openwrt-build/meson-logs/meson-log.txt
```

- https://downloads.openwrt.org/snapshots/faillogs/aarch64_cortex-a53/packages/libqrtr-glib/compile.txt
```
Package libqrtr-glib is missing dependencies for the following libraries:
libgio-2.0.so.0
libglib-2.0.so.0
libgobject-2.0.so.0
```

- https://downloads.openwrt.org/snapshots/faillogs/aarch64_cortex-a53/packages/bluez/compile.txt
```
Package bluez-utils-btmon is missing dependencies for the following libraries:
libglib-2.0.so.0
```

**Unfortunately, this doesn't seem to be ready, so I'm reverting it. It brings very little value at the moment. Once it's ready, feel free to resubmit.**

BKPepe commented:
Reverted commits:
- https://github.com/openwrt/packages/commit/9e3d0983825346b72c8fd66752fa1726b33aee13
- https://github.com/openwrt/packages/commit/5006b17f60a899cb67e892bbb713cc54f05cab4c

pnd10 commented:
That's very odd - libqrtr-glib builds for me on the ATH79 target. (The other two don't but nothing to do with glib2). I will do more investigation and retry. This was designed to be a non-breaking change, as the original glib2 package is built depending on the others, so all existing dependencies should have been OK.

pprindeville commented:
@pnd10 Thanks for digging into that.  If you need help, holler.

pnd10 commented:
> @pnd10 Thanks for digging into that. If you need help, holler.

One thing that would be useful is to know where I can get the configs that the buildbots are using. As I said, at least one of the packages builds for me, but not on the buildbot.

pnd10 commented:
Now I'm really confused. I've just pulled from main, pulled packages from main, put these two commits on top, configured for "Microchip Switches" which is an aarch64_cortex-a53_musl target and then built libqmi, libqrtr-glib and bluez all without any errors. I don't know what the buildbot is doing, or how to proceed.
